kore/src/Kore/Step/BaseStep.hs

```diff
@@ -890,3 +863,526 @@ unwrapPredicateVariables
+
+ -}
+checkSubstitutionCoverage
+    ::  ( MetaOrObject level
+        , Monad m
+        , SortedVariable variable
+        , Ord     (variable level)
+        , Show    (variable level)
+        , Unparse (variable level)
+        )
+    => ExpandedPattern level (StepperVariable variable)
+    -- ^ Initial configuration
+    -> UnifiedRule (StepperVariable variable)
+    -- ^ Unified rule before instantiation
+    -> ExpandedPattern level (StepperVariable variable)
+    -- ^ Configuration after applying rule
+    -> m (ExpandedPattern level variable)
+checkSubstitutionCoverage initial unified final = do
+    (Monad.unless checkPass . error . show . Pretty.vsep)
+        [ "While applying axiom:"
+        , Pretty.indent 4 (Pretty.pretty axiom)
+        , "from the initial configuration:"
+        , Pretty.indent 4 (unparse initial)
+        , "to the final configuration:"
+        , Pretty.indent 4 (unparse final)
+        , "Failed substitution coverage check!"
+        , "Expected substitution to cover all variables:"
+        , (Pretty.indent 4 . Pretty.sep)
+            (unparse <$> Set.toAscList leftAxiomVariables)
+        , "in the left-hand side of the axiom."
+        ]
+    return (unwrapVariables final)
+  where
+    checkPass = isCoveringSubstitution || isInitialSymbolic
+    Predicated { term = axiom } = unified
+    leftAxiomVariables =
+        Pattern.freeVariables leftAxiom
+      where
+        RulePattern { left = leftAxiom } = axiom
+    Predicated { substitution } = final
+    substitutionVariables = Map.keysSet (Substitution.toMap substitution)
+    isCoveringSubstitution =
+        Set.isSubsetOf leftAxiomVariables substitutionVariables
+    isInitialSymbolic =
+        (not . Set.null)
+            (Set.filter isConfigurationVariable substitutionVariables)
+
+{- | Remove axiom variables from the substitution and unwrap all variables.
+ -}
+unwrapVariables
+    :: Ord (variable level)
+    => ExpandedPattern level (StepperVariable variable)
+    -> ExpandedPattern level variable
+unwrapVariables config@Predicated { substitution } =
+    ExpandedPattern.mapVariables unwrapStepperVariable
+        config { substitution = substitution' }
+  where
+    substitution' = Substitution.filter isConfigurationVariable substitution
+
+{- | Negate the coverage of many rules to form the /remainders/.
+
+The /remainders/ are the parts of the initial configuration that are not matched
+by any applied rule.
+
+See also: 'unificationCoverage'
+
+ -}
+negateCoverage
+    ::  ( Ord     (variable Object)
+        , Show    (variable Object)
+        , Unparse (variable Object)
+        , SortedVariable variable
+        )
+    => MultiOr (MultiAnd (Predicate Object variable))
+    -> MultiOr (Predicate Object variable)
+negateCoverage = negateCoverageWorker . Foldable.toList
+  where
+    negateCoverageWorker [] = MultiOr.singleton Predicate.makeTruePredicate
+    negateCoverageWorker (x : xs) =
```

Comment:
Would this be more clear if it was build with one of the three `MultiOr.*crossProduct*` functions?

Reply:
Do you mean like this?
```.hs
    negateCoverageWorker [] = MultiOr.singleton Predicate.makeTruePredicate
    negateCoverageWorker (x : xs) =
        MultiOr.crossProductGeneric
            Predicate.makeAndPredicate
            (mkNotMultiAnd x)
            (negateCoverageWorker xs)
```

Reply:
I meant something like
```hs
negateCoverageWorker xs =
    fmap
        Predicate.makeMultipleAndPredicate 
        (MultiOr.fullCrossProduct
            (map makeNotOnEachPieceOfAnd xs)
        )
```